nsot/models.py

```diff
@@ -1718,6 +1719,186 @@ def to_dict(self):
+    )
+    device = models.ForeignKey(
+        Device, db_index=True, null=False, related_name='protocols',
+        help_text='Device that this protocol is running on',
+    )
+    interface = models.ForeignKey(
+        Interface, db_index=True, blank=True, null=True,
+        related_name='protocols',
+        help_text=(
+            'Interface this protocol is running on. Either interface or'
+            ' circuit must be populated.'
+        ),
+    )
+    circuit = models.ForeignKey(
+        Circuit, db_index=True, blank=True, null=True,
+        related_name='protocols',
+        help_text='Circuit that this protocol is running over.',
+    )
+    auth_string = models.CharField(
+        max_length=255, default='', blank=True,
+        help_text='Authentication string (such as MD5 sum)',
+    )
+    description = models.CharField(
+        max_length=255, default='', blank=True,
+        help_text='Description for this Protocol'
+    )
+
+    def __unicode__(self):
+        return u'%s over %s' % (self.get_type_display(), self.circuit)
+
+    class Meta:
+        ordering = ('device', )
+
+    def local_interface(self):
+        """
+        Returns the local interface for this Protocol, either the set interface
+        or the local interface of circuit
+        """
+        if self.interface:
+            return self.interface
+        elif self.circuit is None:
+            return None
+
+        for endpoint in (self.circuit.endpoint_a, self.circuit.endpoint_z):
+            if endpoint.device.id == self.device.id:
+                return endpoint
+
+        return None
+
+    def remote_interface(self):
+        """
+        Returns the remote device's interface attached to the circuit
+        """
+        if self.circuit is None:
+            return None
+
+        for endpoint in (self.circuit.endpoint_a, self.circuit.endpoint_z):
+            if endpoint.device.id != self.device.id:
+                return endpoint
+
+        return None
+
+    def clean_site(self, value):
+        if value is None:
+            return self.device.site_id
+
+        return value
+
+    def clean_circuit(self, value):
+        """ Ensure at least one endpoint on the circuit is on this device """
+        if value and self.local_interface() is None:
+            raise exc.ValidationError({
+                'circuit': (
+                    'At least one endpoint of the circuit must match the '
+                    'device'
+                )
+            })
+
+        return value
+
+    def clean_interface(self, value):
+        """
+        Ensure that the interface is bound to the same device this Protocol is
+        bound to
+        """
+        if value and value.device != self.device:
+            raise exc.ValidationError({
+                'interface': (
+                    'The interface must be on the same device that this'
+                    ' Protocol is on'
+                )
+            })
+
```

Comment:
Wouldn't the above check pass if the protocol type was set to BGP and a valid `interface` that belonged to the device was specified instead of a `circuit`?